riscv/src/continuations/memory_merkle_tree.rs

```diff
@@ -0,0 +1,274 @@
+    MerkleTree<T, MERKLE_TREE_DEPTH, WORDS_PER_PAGE>
+{
+    /// Build a new Merkle tree starting from an all-zero memory.
+    pub fn new() -> Self {
+        let zero_page = [T::zero(); WORDS_PER_PAGE];
+        let mut hash = Self::hash_page(&zero_page);
+        let mut hashes = vec![vec![]; MERKLE_TREE_DEPTH + 1];
+
+        assert!(usize::BITS >= MERKLE_TREE_DEPTH as u32);
+        for level in (0..=MERKLE_TREE_DEPTH).rev() {
+            hashes[level] = vec![hash; 1 << level];
+            hash = hash_cap0(&hash, &hash);
+        }
+
+        assert_eq!(hashes[0].len(), 1);
+
+        let data = vec![zero_page; 1 << MERKLE_TREE_DEPTH];
+
+        Self { hashes, data }
+    }
+
+    /// Computes the linearly iterated hash of a single page
+    fn hash_page(page: &[T; WORDS_PER_PAGE]) -> [T; 4] {
+        let mut hash = [T::zero(); 4];
+        for chunk in page.chunks_exact(4) {
+            hash = hash_cap0(&hash, chunk.try_into().unwrap());
+        }
+        hash
+    }
+
+    /// Applies updates, given an iterator of (memory address, value) pairs.
+    /// Memory addresses are assumed to be word-aligned.
+    pub fn update(&mut self, updates: impl Iterator<Item = (u32, u32)>) {
+        // Organize by page
+        let mut updates_by_page: BTreeMap<usize, Vec<(usize, u32)>> = BTreeMap::new();
+        for (addr, value) in updates {
+            assert!(addr % BYTES_PER_WORD as u32 == 0);
+            let word_index = addr as usize / BYTES_PER_WORD;
+            let page_index = word_index / WORDS_PER_PAGE;
+            let index_within_page = word_index - page_index * WORDS_PER_PAGE;
+            updates_by_page
+                .entry(page_index)
+                .or_default()
+                .push((index_within_page, value));
+        }
+
+        // Update each page
+        for (page_index, updates) in updates_by_page {
+            let page = &mut self.data[page_index];
+            for (index, value) in updates {
+                page[index] = T::from(value);
+            }
+            self.update_hashes(page_index)
+        }
+    }
+
+    /// Updates the hashes of a page and all its ancestors.
+    fn update_hashes(&mut self, page_index: usize) {
+        self.hashes[MERKLE_TREE_DEPTH][page_index] = Self::hash_page(&self.data[page_index]);
+        for (level, index) in self.iter_path(page_index).skip(1) {
+            self.hashes[level][index] = hash_cap0(
+                &self.hashes[level + 1][index * 2],
+                &self.hashes[level + 1][index * 2 + 1],
+            );
+        }
+    }
+
+    #[allow(dead_code)]
+    /// Returns the root hash of the Merkle tree.
+    pub fn root_hash(&self) -> &[T; 4] {
+        &self.hashes[0][0]
+    }
+
+    /// Returns the data and Merkle proof for a given page.
+    pub fn get(&self, page_index: usize) -> (&[T; WORDS_PER_PAGE], Vec<&[T; 4]>) {
+        let mut proof = vec![];
+        for (level, index) in self.iter_path(page_index).take(MERKLE_TREE_DEPTH) {
```

Comment:
lovely!

Reply:
Ah wait, why `take(MERKLE_TREE_DEPTH)`?

Reply:
Because we want to skip the root level (there, we don't have any sibling).

I think I should probably increase `MERKLE_TREE_DEPTH` by one in the next PR :D There are levels from 0 (inclusive) to 22 (inclusive), so the tree depth should be 23, not 22... Right now, it's really the maximum level, not the depth.

Reply:
Oh! Then maybe `.rev().skip(1).rev()` would be easier to understand?